sstables/sstables.cc

```diff
@@ -3320,8 +3320,63 @@ gc_clock::time_point sstable::get_gc_before_for_fully_expire(const gc_clock::tim
     return res.knows_entire_range ? res.min_gc_before : gc_clock::time_point::min();
 }
 
+future<> remove_table_directory_if_has_no_snapshots(fs::path table_dir) {
+    // Be paranoid about risky paths
+    if (table_dir == "" || table_dir == "/") {
+        on_internal_error_noexcept(sstlog, format("Invalid table directory for removal: {}", table_dir));
+        abort();
+    }
+
+    // Return error code, 0 is success
+    auto remove_dir = [] (fs::path dir, bool recursive) -> future<int> {
+        try {
+            co_await (recursive ? recursive_remove_directory(dir) : remove_file(dir.native()));
+            co_return 0;
+        } catch (...) {
+            auto ex = std::current_exception();
+            int error_code = -1;
+            try {
+                std::rethrow_exception(ex);
+            } catch (const std::system_error& e) {
+                error_code = e.code().value();
+                if (error_code == ENOENT) {
+                    // Ignore missing directories
+                    co_return 0;
+                }
+                if (error_code == EEXIST && !recursive) {
+                    // Just return failure if directory notempty
+                    co_return error_code;
+                }
+            } catch (...) {
+            }
+            sstlog.warn("Could not remove table directory {}: {}. Ignored.", dir, ex);
+            co_return error_code;
+        }
+    };
+
+    int error = 0;
+    for (auto subdir : sstables::table_subdirectories) {
+        // Remove the snapshot directory only if empty
+        // while other subdirectories are removed recusresively.
```

Comment:
I'm still curious why remove the other subdirectories recursively. What can be there? Shouldn't the other stuff be removed when you stop the view building, stop the repair, and so on - and we expect all of them to be empty too?

Reply:
We do expect them to be empty on the error-free path.
I just don't want to depend on that when removing the table, since it will make the remove path more fragile, and needlessly so.